### ISSUE TITLE
Note to myself: make uax License compatibe with Gio UI

### ISSUE THREAD

**Issue by npillmayer** [npillmayer/uax#2]
I've take a look at Gio and have been impressed. Then I did some experiments by coding little samples and have been impressed even more.

Following the discussions about licensing for gu-text, I'll change my license to be fully compatible with Gio.

**Comment by npillmayer** [comment]
Will have to remove bidi, as it pulls in GorGo and that will not be compatible for now. Besides, bidi is super-experimental and should be put on a feature-branch anyway.

Version will have to move up without changing dependency to npillmayer/schuko for now.

**Comment by npillmayer** [comment]
https://choosealicense.com/licenses/

Reading up on licenses and Unlicense in particular. Seems to be a contested topic with the advent of Unlicense. Even read about it not being legally valid Germany, which would be a problem for me (being German), obviously.

Gio/Elias Naur, as well as Raph Levien, use a dual licensing scheme to pacifiy the lawyers. Not sure this is possible with BSD-3, but I'll find out.

**Comment by npillmayer** [comment]
https://www.synopsys.com/blogs/software-security/software-licensing-decisions-consider-dual-licensing/

> This case exemplifies the caution licensees must exercise and the questions a licensee must ask. Licensees must first ask whether a dual licensing model exists. If so, a licensee must then weigh the two license options and ask which set of underlying obligations inherent in the particular licensing option permits the licensee to use or incorporate the software in a way that fits the licensee’s intended purpose. After selecting the licensing option that best aligns with the intended use, a licensee must be sure to ask if any future license obligations exist. Licensees subject to future license obligations should then work to implement a sensible plan to monitor and ensure license compliance.
>
> A licensee must remember that just because software may be freely available, it does not mean there are no obligations. Unknowingly or negligently using software subject to a restrictive open source license such as the GPL may create unintended consequences for licensees who fail to understand the underlying obligations. A licensee prepared to ask the right questions is the most equipped to avoid compliance issue and avoid costly compliance failures.

Seems to cover my intention: allow commercial forks under BSD-3, allow usage by Gio under Unlicense

**Comment by npillmayer** [comment]
The SQLite approach:

> Contributed Code
>
> In order to keep SQLite completely free and unencumbered by copyright, the project does not accept patches. If you would like to suggest a change and you include a patch as a proof-of-concept, that would be great. However, please do not be offended if we rewrite your patch from scratch.

I guess the Contributor-signing-Approach is to be preferred for me :-)

**Comment by npillmayer** [comment]
Info: uax no longer depends on GorGo. Obviously that has been a stale dependency on experimental code I planned to remove anyway.

Cords depends on uax/bidi. Therefore not sure how to proceed with forking out bidi

**Comment by npillmayer** [comment]
Next complication: uax includes a copy of an internal Go dir. It is used for generating tables. **This code can certainly not be made unlicensed**.

Option A: move this code someplace else, as it is used for table generation only.

Option B: implement the UCD file parser from scratch.

I don't like option A. Need to find a spare evening for Option B.

**Comment by npillmayer** [comment]
Coded the **UAX file format parser** from scratch. Luckily I'd finished may NestedText parser a couple of days ago and could re-use the scanner with minor changes. I also took this as an occasion to include all the table generators into the module (in 'internal' packages). However, not included are the Unicode source files.

**Comment by npillmayer** [comment]
Info to @whereswaldon: The complete npillmayer/uax repo has now switched to a dual licensing scheme: UNLICENSE or 3-Clause-BSD.

Every line of code (and every bug) is now written by myself from scratch. Dual licensing will take effect for all future versions > v0.1.0

Due to a confusion on my part the license-switch lives on a branch intended to upgrade to a refactored npillmayer/schuko and therefore will for now depend on an alpha version of schuko (v0.2.0-alpha). This will change soon and I'll move both packages to v0.2.0. Meanwhile it will probably be wise to stay at 0.1.0 for the time being. If you encounter any problems due to the ongoing logging refactoring, **please do not hesitate to file an issue**!

I moved out the *experimental bidi module* onto a feature branch. I'm currently working on fonts and have postponed the bidi topic for a couple of weeks. If ui-text acitivities or Gio activities are nearing a point where bidi comes into play, please be so kind as to reach out to me for support.

**Comment by whereswaldon** [comment]
@npillmayer Wow! I appreciate you reading through go-text/shaping#6, your subsequent research, and the effort you've made to convert all of this. It's very generous of you to make your hard work available in the public domain.

For bidi, I'm not certain when it will come into play, but I will certainly reach out when it does. I'm focused on solving the simpler "single-run in either horizontal direction" case first, and integrating all of that work back into Gio's text shaping stack.

Not to complicate your life, but I wonder if the [CC0 License](https://creativecommons.org/share-your-work/public-domain/cc0/) has the same problems in German law as the Unlicense. It's still compatible with Gio/Unlicense in the sense that it places your work in the public domain (or as close as is possible), but might be better in your jurisdiction. That being said, I'm happy either way.

Thanks again for all of your work here. It's really impressive.

**Comment by npillmayer** [comment]
No worries, this supposed issue with German law is easily resolved. It's more of a strawman argument against public domain licenses, I suspect.

I've been following [Raph Levien](https://raphlinus.github.io/) quite some time, since the early days of the Xi editor. Recently I discovered the link from him to Elias Naur and further along to Gio and yourself. As I said, I'm not really into GUIs, but I find the depth and challenges around the topics of immediate UI and GPU rendering pipelines fascinating. I'll always cheer for anyone doing stuff in fresh and ingenious ways!

Thank you for taking the time to give feedback! Closing this issue.
